File: hw2.py
def hw2(input, output):
    f_in = open(input, 'r')
    f_out = open(output, 'w')
    stack = []
    for line in f_in:
        splitted = line.split()
        command = splitted[0]
        if command == "quit":
            break
        else:
            ops(command, splitted, stack)
    stack.reverse()
    for b in stack:
        b = str(b)
        f_out.write(b)
        f_out.write('\n')

    f_in.close()
    f_out.close()


def ops(operator, element, stack):
    # push
    if operator == "push":
        try:
            val = int(element[1])
            stack.append(val)
        except ValueError:
            stack.append(":error:")
    # pop
    elif operator == "pop":
        if not stack:
            stack.append(":error:")
        else:
            stack.pop()
    elif operator == ":true:":
        stack.append(":true:")
    elif operator == ":false:":
        stack.append(":false:")
    elif operator == ":error:":
        stack.append(":error:")
    elif operator == "add":
        if len(stack) < 2:
            stack.append(":error:")
        else:
            x = stack.pop()
            y = stack.pop()
            add_check(x, y, stack)
    elif operator == "sub":
        if len(stack) < 2:
            stack.append(":error:")
        else:
            x = stack.pop()
            y = stack.pop()
            sub_check(x, y, stack)
    elif operator == "mul":
        if len(stack) < 2:
            stack.append(":error:")
        else:
            x = stack.pop()
            y = stack.pop()
            mul_check(x, y, stack)
    elif operator == "div":
        if len(stack) < 2:
            stack.append(":error:")
        else:
            x = stack.pop()
            y = stack.pop()
            div_check(x, y, stack)
    elif operator == "rem":
        if len(stack) < 2:
            stack.append(":error:")
        else:
            x = stack.pop()
            y = stack.pop()
            rem_check(x, y, stack)
    elif operator == "neg":
        if len(stack) < 1:
            stack.append(":error:")
        else:
            x = stack.pop()
            if x == 0:
                stack.append(x)
            else:
                try:
                    x = int(x)
                    stack.append(-x)
                except ValueError:
                    stack.append(x)
                    stack.append(":error:")
    elif operator == "swap":
        if len(stack) < 2:
            stack.append(":error:")
        else:
            x = stack.pop()
            y = stack.pop()
            stack.append(x)
            stack.append(y)
    else:
        print("Command not recognized")


def add_check(a, b, stack):
    try:
        stack.append(int(a)+int(b))
    except ValueError:
        stack.append(b)
        stack.append(a)
        stack.append(":error:")


def sub_check(a, b, stack):
    try:
        stack.append(int(b)-int(a))
    except ValueError:
        stack.append(b)
        stack.append(a)
        stack.append(":error:")


def mul_check(a, b, stack):
    try:
        stack.append(int(a)*int(b))
    except ValueError:
        stack.append(b)
        stack.append(a)
        stack.append(":error:")


def div_check(a, b, stack):
    try:
        stack.append(int(b)//int(a))
    except ValueError:
        stack.append(b)
        stack.append(a)
        stack.append(":error:")
    except ZeroDivisionError:
        stack.append(b)
        stack.append(a)
        stack.append(":error:")


def rem_check(a, b, stack):
    if b == 0:
        stack.append(b)
        stack.append(a)
        stack.append(":error:")
    else:
        try:
            stack.append(int(b) % int(a))
        except ValueError:
            stack.append(b)
            stack.append(a)
            stack.append(":error:")

File: cse305/hw2.java
/**
 * Created by Dean on 3/3/2016.
 */
import java.io.*;
import java.util.Stack;

//TODO: Check rem and div functions - make sure they're working properly
//TODO: Fix file writing
//TODO: Finish other ops - check edge cases

public class hw2 {
    public static void hw2(String inFile, String outFile){
        try {
            FileReader fr = new FileReader(inFile);
            BufferedReader bReader = new BufferedReader(fr);
            FileWriter fw = new FileWriter(outFile);
            BufferedWriter bWriter = new BufferedWriter(fw);
            Stack<String> elemStack = new Stack<>();
            String buffer;
            while((buffer = bReader.readLine()) != null){
                //read each line do operations
                String[] splitter = buffer.split("\\s+");
                if(splitter[0].equals("quit")) break;
                ops_Check(splitter, elemStack);
                //bWriter.write(elemStack);
                //System.out.println(elemStack);
            }
            System.out.println(elemStack); //need to reverse?
            System.out.println("size is : " + elemStack.size());
            while(elemStack.isEmpty() != true){
                //System.out.println(elemStack.pop());
                bWriter.write(elemStack.pop() + '\n');
            }
            bWriter.close();
            bReader.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

    }
    public static void ops_Check(String[] op, Stack<String> tarStack){

        switch(op[0]) {
            case "push":
                try {
                    tarStack.push(String.valueOf(Integer.parseInt(op[1])));
                } catch (NumberFormatException nonInt) {
                    tarStack.push(":error:");
                }
                break;
            case "pop":
                if (tarStack.isEmpty()) {
                    tarStack.push(":error:");
                } else {
                    tarStack.pop();
                }
                break;
            case ":true:":
                tarStack.push(":true:");
                break;
            case ":false:":
                tarStack.push(":false:");
                break;
            case ":error:":
                tarStack.push(":error:");
                break;
            case "add":
                if (tarStack.size() < 2) {
                    tarStack.push(":error:");
                } else {
                    String x = tarStack.pop();
                    String y = tarStack.pop();
                    try {
                        tarStack.push(String.valueOf(Integer.parseInt(x) + Integer.parseInt(y)));
                    } catch (NumberFormatException nonInt) {
                        tarStack.push(y);
                        tarStack.push(x);
                        tarStack.push(":error:");
                    }
                }
                break;
            case "sub":
                if (tarStack.size() < 2) {
                    tarStack.push(":error:");
                } else {
                    String x = tarStack.pop();
                    String y = tarStack.pop();
                    try {
                        tarStack.push(String.valueOf(Integer.parseInt(y) - Integer.parseInt(x)));
                    } catch (NumberFormatException nonInt) {
                        tarStack.push(y);
                        tarStack.push(x);
                        tarStack.push(":error:");
                    }
                }
                break;
            case "mul":
                if (tarStack.size() < 2) {
                    tarStack.push(":error:");
                } else {
                    String x = tarStack.pop();
                    String y = tarStack.pop();
                    try {
                        tarStack.push(String.valueOf(Integer.parseInt(y) * Integer.parseInt(x)));
                    } catch (NumberFormatException nonInt) {
                        tarStack.push(y);
                        tarStack.push(x);
                        tarStack.push(":error:");
                    }
                }
                break;
            case "div":
                if (tarStack.size() < 2) {
                    tarStack.push(":error:");
                } else {
                    String x = tarStack.pop();
                    String y = tarStack.pop();
                    try {
                        tarStack.push(String.valueOf(Integer.parseInt(y) / Integer.parseInt(x)));
                    } catch (NumberFormatException | ArithmeticException  badInt) {
                        tarStack.push(y);
                        tarStack.push(x);
                        tarStack.push(":error:");
                    }
                }
                break;
            case "rem":
                if (tarStack.size() < 2) {
                    tarStack.push(":error:");
                }else{
                    String x = tarStack.pop();
                    String y = tarStack.pop();
                    try{
                        tarStack.push(String.valueOf(Integer.parseInt(y) % Integer.parseInt(x)));
                    } catch (NumberFormatException | ArithmeticException badInt) {
                        tarStack.push(y);
                        tarStack.push(x);
                        tarStack.push(":error:");
                    }
                }
                break;
            case "neg":
                if(tarStack.isEmpty()){
                    tarStack.push(":error:");
                } else {
                    String negString = tarStack.pop();
                    try {
                        int negative = Integer.parseInt(negString);
                        if(negative == 0){
                            tarStack.push(String.valueOf(negative));
                        }else{
                            tarStack.push(String.valueOf(-negative));
                        }
                    } catch(NumberFormatException e) {
                        tarStack.push(negString);
                        tarStack.push(":error:");
                    }
                }
                break;
            case "swap":
                if(tarStack.size() < 2){
                    tarStack.push(":error:");
                } else {
                    String x = tarStack.pop();
                    String y = tarStack.pop();
                    tarStack.push(x);
                    tarStack.push(y);
                }
                break;
        }
    }
}
